### PR TITLE
Introduce before and after steps for a task.

### PR DESCRIPTION
**Pull request by jenpet**
Hey guys, I fiddled around with some `before` and `after` functionality I could use within robo. Hopefully this finds a way into master. Let me know what you think, will update the docs after we're all happy with it :-)

A task is now able to have several steps which are executed before and after its own execution. The config syntax for this is the same as for the task itself: 'command', 'script' and 'exec' are supported. A task and all of the preceeding steps will still be executed even when the before steps are failing.

Running a task now returns a list of errors which are logged according to their occurrence for easier troubleshooting.

### CODE REVIEW

**Comment by jenpet** [comment]
@yields *ping* 😊

File: interpolation/interpolation.go [hunk 1]
```diff
@@ -91,13 +93,38 @@ func Tasks(tasks map[string]*task.Task, data map[string]interface{}) error {
...
+			return err
+		}
+		if err := Optionals("after", task.After, data); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func Optionals(id string, rs []*task.Runnable, data map[string]interface{}) error {
```
Comment (yields) [inline interpolation/interpolation.go:115]: Can we add docs to this one? 🙏 Just to match the rest of the codebase.

File: task/task.go [hunk 2]
```diff
@@ -28,32 +28,66 @@ type Task struct {
...
-	if t.Command != "" {
-		return t.RunCommand(args)
+	return errs
+}
+
+func (t *Task) runTaskOptionals(id string, rs []*Runnable, args []string) error {
+	return RunOptionals(id, t.Name, rs, args, t.LookupPath, t.Env)
+}
+
+type Runnable struct {
```
Comment (yields) [inline task/task.go:62]: ditto

File: task/task.go [hunk 3]
```diff
@@ -28,32 +28,66 @@ type Task struct {
...
+	Command string
+	Script  string
+	Exec    string
+}
+
+func (r *Runnable) Run(lookupPath string, args []string, env []string) error {
```
Comment (yields) [inline task/task.go:68]: ditto

File: task/task.go [hunk 4]
```diff
@@ -129,3 +163,12 @@ func merge(a, b []string) []string {
 
 	return ret
 }
+
+func RunOptionals(id string, parent string, rs []*Runnable, args []string, lookupPath string, envs []string) error {
```
Comment (yields) [inline task/task.go:167]: ditto

**Review by yields — CHANGES_REQUESTED** [review]
LGTM, seems like it just needs some docs to match the rest of the codebase.

**Comment by jenpet** [comment]
@yields updated the code documentation and the readme accordingly. If you are happy with it, I will squash the commits into a single reasonable commit.

**Review by yields — APPROVED** [review]
Awesome! -- thanks for the extra docs :)

**Pull request merged**